File: apps/backend/fastapi/app/schemas/cv.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from dlib.cv.schema import PatchSuggestion, StructuredBlock


class DocumentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    description: str | None
    owner_id: str
    root_version_id: str | None
    created_at: datetime
    updated_at: datetime
    versions: list["VersionResponse"] = Field(default_factory=list)


class VersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    branch_name: str
    version_label: str | None
    parent_version_id: str | None
    structured_blocks: list[StructuredBlock] | None = None
    artifact_docx_key: str | None = None
    preview_html_key: str | None = None
    created_at: datetime
    updated_at: datetime
    patches: list["PatchResponse"] = Field(default_factory=list)


class PatchResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    target_path: str
    operation: str
    old_value: str | None = None
    new_value: str | None = None
    metadata_json: dict[str, Any] | None = None
    created_at: datetime


class DocumentListResponse(BaseModel):
    items: list[DocumentResponse]


class DocumentCreateResult(BaseModel):
    document: DocumentResponse


class BranchCreateRequest(BaseModel):
    parent_version_id: str
    branch_name: str
    version_label: str | None = None
    patches: list[dict[str, Any]] = Field(default_factory=list)


class SubmissionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    version_id: str
    company_name: str
    role_title: str
    job_url: str | None = None
    job_description: str | None = None
    status: str
    created_at: datetime
    suggestions: list["SuggestionResponse"] = Field(default_factory=list)


class SubmissionCreateRequest(BaseModel):
    version_id: str
    company_name: str
    role_title: str
    job_url: str | None = None
    job_description: str | None = None


class SuggestionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    target_path: str
    operation: str
    proposed_text: str | None
    rationale: str | None
    accepted: bool | None
    metadata_json: dict[str, Any] | None = None


class AiSuggestionRequest(BaseModel):
    job_description: str
    focus_keywords: list[str] = Field(default_factory=list)


class PublishRequest(BaseModel):
    version_id: str | None = None
    submission_id: str | None = None
    slug: str | None = None


class PublicAssetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    slug: str
    artifact_key: str
    is_public: bool
    created_at: datetime
    version_id: str | None = None
    submission_id: str | None = None
    url: str | None = None


class PublicAssetLookupResponse(BaseModel):
    asset: PublicAssetResponse

File: apps/webapp/src/types/cv.ts
// Core data types for Resume Branches system

export interface CVDocument {
  id: string;
  ownerId: string;
  title: string;
  rootVersionId: string;
  createdAt: string;
  updatedAt: string;
}

export interface CVVersion {
  id: string;
  documentId: string;
  parentVersionId?: string;
  branchName?: string;
  versionLabel: string;
  artifactDocxKey: string;
  previewHtmlKey?: string;
  createdBy: string;
  createdAt: string;
  isPublic: boolean;
}

export interface CVPatch {
  id: string;
  versionId: string;
  patchType: 'text_replace' | 'reorder' | 'add' | 'remove';
  targetPath: string; // e.g., "experience[1].bullets[2]" or "summary.paragraph_1"
  operation: string;
  oldValue?: string;
  newValue?: string;
  metadata?: Record<string, unknown>;
}

export interface Specialization {
  id: string;
  documentId: string;
  name: string;
  basedOnVersionId: string;
  description?: string;
  color?: string; // For UI visualization
}

export interface Submission {
  id: string;
  versionId: string;
  companyName: string;
  roleTitle: string;
  jobUrl?: string;
  jobDescription?: string;
  status: 'draft' | 'submitted' | 'rejected' | 'interviewing' | 'offer' | 'closed';
  publicAssetId?: string;
  createdAt: string;
  updatedAt: string;
}

export interface PublicAsset {
  id: string;
  submissionId?: string;
  versionId?: string;
  slug: string;
  artifactKey: string;
  isPublic: boolean;
  expiresAt?: string;
  viewCount: number;
  url?: string | null;
  paperlessShareUrl?: string | null;
}

export interface AISuggestion {
  id: string;
  submissionId: string;
  targetPath: string;
  suggestionType: 'strengthen_verb' | 'add_keyword' | 'reorder' | 'shorten' | 'expand';
  proposedText: string;
  rationale: string;
  accepted?: boolean;
  appliedAt?: string;
}

// UI-specific types
export interface CVTreeNode {
  id: string;
  label: string;
  type: 'root' | 'branch' | 'submission';
  versionId: string;
  parentId?: string;
  children: CVTreeNode[];
  metadata?: {
    companyName?: string;
    roleTitle?: string;
    status?: string;
    isPublic?: boolean;
    lastModified?: string;
  };
}

export interface PatchDiff {
  path: string;
  type: 'added' | 'removed' | 'changed';
  oldValue?: string;
  newValue?: string;
  context?: string;
}

export interface CVStructure {
  header: {
    name: string;
    title: string;
    contact: string[];
  };
  summary?: {
    paragraphs: string[];
  };
  experience: Array<{
    company: string;
    title: string;
    duration: string;
    bullets: string[];
  }>;
  education: Array<{
    institution: string;
    degree: string;
    duration: string;
    details?: string[];
  }>;
  skills: {
    [category: string]: string[];
  };
  sections?: Array<{
    title: string;
    content: string[] | Record<string, unknown>;
  }>;
}